lean4.html.markdown

```diff
@@ -0,0 +1,587 @@
+-/
+theorem contrapositive_forward : (p → q) → (¬q → ¬p) := by
+  intro hpq
+  intro
+  intro hp
+  specialize hpq hp
+  contradiction
+
+/-
+The reverse requires classical logic.
+
+Here, we are required to construct a `q` given values of following types:
+
+  - `(q → False) → (p → False)`.
+  - `p`.
+
+This is impossible without using the law of excluded middle.
+-/
+theorem contrapositive_reverse' : (¬q → ¬p) → (p → q) :=
+  fun hnqnp =>
+  Classical.byCases
+    (fun hq  => fun  _ => hq)
+    (fun hnq => fun hp => absurd hp (hnqnp hnq))
+/-
+Law of excluded middle tells us that we will have a `q` or a `q → False`. In the
+first case, it is trivial to construct a `q`, we already have it. In the second
+case, we give the `q → False` to obtain a `p → False`.  Then, we use the fact
+(in constructive logic) that given `p` and `p → False`, we can construct
+`False`. Once, we have `False`, we can construct anything, and specifically `q`.
+-/
+
+/-
+Same proof, using tactics.
+-/
+theorem contrapositive_reverse : (¬q → ¬p) → (p → q) := by
+  intro hnqnp
+  intro hp
+  have emq := Classical.em q
+  cases emq
+  case inl _ => assumption
+  case inr h => specialize hnqnp h ; contradiction
+
+/-
+To illustrate how we can define an work with axiomatic systems. Here is a
+definition of Groups and some proofs directly translated from "Topics in
+Algebra" by Herstein, Second edition.
+-/
+
+/-
+A class defines a set of axioms
+-/
+class Group (G : Type u) where
+  op : G → G → G
+  assoc : ∀ a b c : G, op (op a b) c = op a (op b c)
+  e : G
+  identity: ∀ a : G, op a e = a ∧ op e a = a
+  inverse: ∀ a : G, ∃ b : G, op a b = e ∧ op b a = e
+
+/-
+Let us introduce some notation to make this convenient.
+-/
+open Group
+infixl:70 " * " => op
+
+/-
+`G` will always stand for a group. Variables `a b c` will be elements of that
+group.
+-/
+variable [Group G] {a b c : G}
+
+def is_identity (e' : G) := ∀ a : G, (a * e' = a ∧ e' * a = a)
+
+/-
+We prove that the identity element is unique.
+-/
+theorem identity_element_unique : ∀ e' : G, is_identity e' → e' = e := by
+  intro e'
+  intro h
+  specialize h e
+  obtain ⟨h1, _⟩ := h
+  have h' := identity e'
+  obtain ⟨_, h2⟩ := h'
+  apply Eq.trans (Eq.symm h2) h1
+/-
+Note that we used the `identity` axiom.
+-/
+
+/-
+Left cancellation. We have to use both `identity` and `inverse` axioms from
+`Group`.
+-/
+theorem left_cancellation : ∀ x y : G, a * x = a * y → x = y := by
+  have h1 := inverse a
+  obtain ⟨ai, a_inv⟩ := h1
+  obtain ⟨_, h2⟩ := a_inv
+  intro x y
+  intro h3
+  calc
+    x = (e : G) * x  := Eq.symm (identity x).right
+    _ = ai * a * x   := by rw [h2]
+    _ = ai * (a * x) := by rw [assoc]
+    _ = ai * (a * y) := by rw [h3]
+    _ = ai * a * y   := by rw [← assoc]
+    _ = (e : G) * y  := by rw [h2]
+    _ = y            := (identity y).right
+
+/-
+Mutually recursive definitions have to be carefully constructed. In particular,
+the hypotheses when recursing should imply well-foundedness.
```

Comment:
This is misleading. Every definition in Lean has to be well-founded, regardless of whether it's mutually defined.